MyMovies/Sources/Features/MovieSearch/ViewModels/MovieSearchViewModel.swift

```diff
@@ -0,0 +1,120 @@
+import Foundation
+import SwiftUI
+
+/// A view model responsible for managing the movie search functionality.
+///
+/// `MovieSearchViewModel` provides the logic and state management for a movie search view.
+/// It interacts with a `Searchable` service to perform asynchronous search operations and updates the state
+/// of the view based on the results or errors.
+///
+/// ### Generic Parameters
+/// - `Service`: A type conforming to the `Searchable` protocol, used to fetch movies based on search queries.
+///
+/// ### States
+/// The `State` enum defines the possible states of the view:
+/// - `idle`: No search has been initiated.
+/// - `loading`: A search operation is in progress.
+/// - `loaded`: Movies have been successfully fetched and are ready to display.
+/// - `failed`: The search operation failed, either due to a network error or no results.
+///
+/// ### Error Handling
+/// The `MovieSearchError` enum defines two possible error cases:
+/// - `networkError`: Indicates a network-related failure during the search.
+/// - `noResults`: Indicates that no movies were found for the given search query.
+///
+/// ### Example Usage
+/// ```swift
+/// let viewModel = MovieSearchViewModel(service: MovieService())
+/// viewModel.searchMovies(with: "Inception")
+/// print(viewModel.state) // Outputs the updated state after performing the search
+/// ```
+@Observable
+class MovieSearchViewModel<Service: Searchable>: ViewModelable {
+    
+    /// Errors that can occur during the movie search process.
+    enum MovieSearchError: Error, Equatable {
+        /// Indicates a network-related error.
+        case networkError
+        /// Indicates that no results were found for the given query.
+        case noResults
+    }
+    
+    /// The various states the view can be in during the search process.
+    enum State: Equatable {
+        /// The view is idle, with no ongoing search operation.
+        case idle
+        /// A search operation is currently in progress.
+        case loading
+        /// Movies have been successfully fetched.
+        case loaded(movies: [Movie])
+        /// The search operation failed with a specific error.
+        case failed(error: MovieSearchError)
+    }
+    
+    /// The current search text entered by the user.
+    var searchText: String = ""
```

Comment:
At first I added the search text to the state, but realized that this didnt create a great user experience as this text could have a value and another state also be appropriate to represent (ie: failed and search text)